src/transformers/modeling_tf_utils.py

```diff
@@ -89,6 +94,82 @@ def __init__(self, config, *inputs, **kwargs):
             )
         # Save config in model
         self.config = config
+    
+    def create_optimizer(self, **kwargs):
+        """
+        Create the training optimizer with its name. Allowed names:
+          - adam: Adam optimizer
+          - adamw: Adam with Weight decay optimizer
+          - adadelta: Adadelta optimizer
+          - rmsprop: Root Mean Square Propogation optimizer
+          - sgd: Stochastic Gradient Descent optimizer
+        
+        Args:
+          **kwargs: keywords arguments. Allowed to be {`scheduler`}.
+            `scheduler` is learning rate scheduler used for the optimizer.
+        """
+        raise NotImplementedError
+
+    def get_loss(self, real, pred):
+        """
+        Custom loss computation function.
+        Note: I still thinking what will be the best way to
+        either customing a loss or using an already existing one
+        """
+        raise NotImplementedError
+
+    def create_checkpoint_manager(self, checkpoint_path, max_to_keep=5, load_model=True):
+        """
+        Create a checkpoint manager in order to be able to make the training
+        fault-tolerant.
+
+        Args:
+          checkpoint_path: the path where the checkpoints will be saved.
+          max_to_keep: the maximum number of checkpoints to keep in the
+            checkpoint path.
+          load_model: if we want to start the training from the latest checkpoint.
+        """
+        raise NotImplementedError
+
+    def create_summary_writer(self, log_path):
+        """
+        Create a summary writer to be able to read the logs in Tensorboard.
+
+        Args:
+          log_path: the path where to write the logs
+        """
+        raise NotImplementedError
+    
+    def fit(self, train_dataset):
+        """
+        Fit method to train the model.
+
+        Args:
+          train_dataset: training dataset.
+        """
+        self.fit(train_dataset)
```

Comment:
I'm not super familiar with TF but isn't there a .fit method out of the box on Keras models?

Reply:
Yes, this is an overwriting of the fit method. The reason is because the custom training loop will not be the same depending if we want distributed on single GPU training. Thus, this overwritten fit will looks like something similar to:
```
if self.distributed:
  loop over distributed_train_step()
else:
  loop over train_step()
```
This is for making the handling of the TF strategies invisible for the user. Thus, making the training distributed or not will be just a boolean in the constructor.